File: IOI '11 P2 - Race (Standard IO).py
"""
TLE, PYTHON IS TOO SLOW, CHECK C++ CODE

https://dmoj.ca/problem/ioi11p2io
Tree all-paths questions
At each node, we store the paths in {sum: min length to achieve sum} for efficient joining
As usual we use 2-sum method to join paths (also consider only using 1 path)

Now we have 2 approaches:
1. Standard tree DP with small to large merging of children and offsetting
   for fast children-to-parent transitions
2. Template centroid decomposition

I'll be using the second method because offsetting to do the transitions is very annoying
TC: O(nlogn)
"""
import sys

inf = 1 << 30
input = sys.stdin.readline
sys.setrecursionlimit(100000)

N, K = map(int, input().split())
graph = [[] for _ in range(N + 1)]
for _ in range(N - 1):
    a, b, c = map(int, input().split())
    graph[a].append((b, c))
    graph[b].append((a, c))

best = inf
size = [0] * (N + 1)
cut = [False] * (N + 1)  # determine is node i is removed


def get_sizes(cur, prev):
    size[cur] = 1
    for adj, _ in graph[cur]:
        if adj == prev or cut[adj]:
            continue
        get_sizes(adj, cur)
        size[cur] += size[adj]


def get_centroid(cur, prev, comp_size):
    for adj, _ in graph[cur]:
        if not cut[adj] and adj != prev and size[adj] > comp_size // 2:
            return get_centroid(adj, cur, comp_size)
    return cur


def get_paths(start, prev):
    stack = [(start, prev, 0, 0)]  # cur, prev, sum, length
    paths = []
    while stack:
        cur, prev, tot, le = stack.pop()
        paths.append((tot, le))
        for adj, val in graph[cur]:
            if adj == prev or cut[adj]:
                continue
            stack.append((adj, cur, tot + val, le + 1))
    return paths


def solve(node):
    global best
    get_sizes(node, -1)
    centroid = get_centroid(node, -1, size[node])
    match = {0: 0}  # include option of only using 1 path

    for adj, val in graph[centroid]:
        if cut[adj]:
            continue
        # include the edge from centroid to adj
        adj_paths = list(map(lambda x: (x[0] + val, x[1] + 1), get_paths(adj, centroid)))
        for tot, le in adj_paths:  # match current subtree's paths with others
            if K - tot in match:
                best = min(best, match[K - tot] + le)  # joined 2 path to sum to K
        for tot, le in adj_paths:  # add current subtree's paths as a matching option
            if tot in match:
                match[tot] = min(match[tot], le)
            else:
                match[tot] = le

    cut[centroid] = True  # remove centroid and recurse to subtrees
    for adj, _ in graph[centroid]:
        if not cut[adj]:
            solve(adj)


solve(0)
print(best if best != inf else -1)

File: IOI '11 P2 - Race (Standard IO).cpp
// check Python code for explanations

#include <bits/stdc++.h>
using namespace std;
const int N = 2e5 + 5;
const int inf = 1 << 30;
vector<pair<int, int>> graph[N];
int best = inf;
int _size[N], K;
bool cut[N];

void get_sizes(int cur, int prev) {
    _size[cur] = 1;
    for(auto &it : graph[cur]) {
        int adj = it.first;
        if(adj == prev || cut[adj]) continue;
        get_sizes(adj, cur);
        _size[cur] += _size[adj];
    }
}

int get_centroid(int cur, int prev, int comp_size) {
    for(auto &it : graph[cur]) {
        int adj = it.first;
        if(!cut[adj] && adj != prev && _size[adj] > comp_size / 2)
            return get_centroid(adj, cur, comp_size);
    }
    return cur;
}

vector<pair<int, int>> get_paths(int start, int prev) {
    vector<pair<int, int>> paths;
    stack<tuple<int, int, int, int>> stk;
    stk.push({start, prev, 0, 0});
    while(!stk.empty()) {
        auto [cur, prev, tot, le] = stk.top(); stk.pop();
        paths.push_back({tot, le});
        for(auto &it : graph[cur]) {
            int adj = it.first, val = it.second;
            if(adj == prev || cut[adj]) continue;
            stk.push({adj, cur, tot + val, le + 1});
        }
    }
    return paths;
}

void solve(int node) {
    get_sizes(node, -1);
    int centroid = get_centroid(node, -1, _size[node]);
    unordered_map<int, int> match = {{0, 0}};
    for(auto &it : graph[centroid]) {
        int adj = it.first, val = it.second;
        if(cut[adj]) continue;
        vector<pair<int, int>> adj_paths;
        for(auto &path : get_paths(adj, centroid))
            adj_paths.push_back({path.first + val, path.second + 1});
        for(auto &path : adj_paths) {
            int tot = path.first, le = path.second;
            if(match.count(K - tot)) best = min(best, match[K - tot] + le);
        }
        for(auto &path : adj_paths) {
            int tot = path.first, le = path.second;
            if(match.count(tot)) match[tot] = min(match[tot], le);
            else match[tot] = le;
        }
    }
    cut[centroid] = true;
    for(auto &it : graph[centroid])
        if(!cut[it.first]) solve(it.first);
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    int N;
    cin >> N >> K;
    for(int i = 1; i < N; i++) {
        int a, b, c;
        cin >> a >> b >> c;
        graph[a].push_back({b, c});
        graph[b].push_back({a, c});
    }
    solve(0);
    cout << (best == inf ? -1 : best) << "\n";
    return 0;
}